File: DevExpress-Demo/VB/TreeListDemo.Wpf/Data/DynamicNodeLoadingBehavior.vb
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Grid
Imports DevExpress.Xpf.Grid.TreeList
Imports System
Imports System.Collections
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows
Imports System.Windows.Media

Namespace TreeListDemo.Data

    Public Class FileSystemChildBehavior
        Inherits Behavior(Of TreeListView)

        Public Property DataProvider As IDataProvider
            Get
                Return CType(GetValue(DataProviderProperty), IDataProvider)
            End Get

            Set(ByVal value As IDataProvider)
                SetValue(DataProviderProperty, value)
            End Set
        End Property

        Public Shared ReadOnly DataProviderProperty As DependencyProperty = DependencyProperty.Register("DataProvider", GetType(IDataProvider), GetType(FileSystemChildBehavior), New PropertyMetadata(Nothing, Sub(s, e) CType(s, FileSystemChildBehavior).OnDataProviderChanged(e)))

        Private Sub OnDataProviderChanged(ByVal e As DependencyPropertyChangedEventArgs)
            If e.NewValue Is Nothing Then Return
            AssociatedObject.ChildNodesSelector = New CustomAsyncChildNodeSelector(DataProvider)
        End Sub
    End Class

    Public Class CustomAsyncChildNodeSelector
        Implements IAsyncChildNodesSelector

        Private ReadOnly dataProvider As IDataProvider

        Public Sub New(ByVal dataProvider As IDataProvider)
            Me.dataProvider = dataProvider
        End Sub

        Public Function HasChildNode(ByVal item As Object, ByVal token As CancellationToken) As Task(Of Boolean) Implements IAsyncChildNodesSelector.HasChildNode
            Return Tasks.Task.Run(Async Function()
                For i As Integer = 0 To 10 - 1
                    token.ThrowIfCancellationRequested()
                    Await Tasks.Task.Delay(25)
                Next

                Return dataProvider.HasChildren(item)
            End Function)
        End Function

        Public Function SelectChildren(ByVal item As Object) As IEnumerable Implements IChildNodesSelector.SelectChildren
            Throw New NotImplementedException()
        End Function

        Public Function SelectChildrenAsync(ByVal item As Object, ByVal token As CancellationToken) As Task(Of IEnumerable) Implements IAsyncChildNodesSelector.SelectChildrenAsync
            Return Tasks.Task.Run(Async Function()
                For i As Integer = 0 To 20 - 1
                    token.ThrowIfCancellationRequested()
                    Await Tasks.Task.Delay(100)
                Next

                Return dataProvider.GetChildren(item)
            End Function)
        End Function
    End Class

    Public Class FileSystemNodeImageSelector
        Inherits TreeListNodeImageSelector

        Public Overrides Function [Select](ByVal rowData As TreeListRowData) As ImageSource
            Dim fsItem = TryCast(rowData.Row, FileSystemItem)
            If fsItem Is Nothing Then Return Nothing
            If Equals(fsItem.ItemType, "Drive") Then
                Return FileSystemImages.DiskImage
            ElseIf Equals(fsItem.ItemType, "Folder") Then
                Return If(rowData.IsExpanded, FileSystemImages.OpenedFolderImage, FileSystemImages.ClosedFolderImage)
            End If

            Return FileSystemImages.FileImage
        End Function
    End Class
End Namespace

File: DevExpress-Demo/CS/TreeListDemo.Wpf/Data/DynamicNodeLoadingBehavior.cs
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Grid;
using DevExpress.Xpf.Grid.TreeList;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Media;

namespace TreeListDemo.Data {

    public class FileSystemChildBehavior : Behavior<TreeListView> {

        public IDataProvider DataProvider {
            get { return (IDataProvider)GetValue(DataProviderProperty); }
            set { SetValue(DataProviderProperty, value); }
        }

        public static readonly DependencyProperty DataProviderProperty =
            DependencyProperty.Register("DataProvider", typeof(IDataProvider), typeof(FileSystemChildBehavior), new PropertyMetadata(null,
                (s, e) => ((FileSystemChildBehavior)s).OnDataProviderChanged(e)));

        void OnDataProviderChanged(DependencyPropertyChangedEventArgs e) {
            if(e.NewValue == null)
                return;
            AssociatedObject.ChildNodesSelector = new CustomAsyncChildNodeSelector(DataProvider);
        }
    }

    public class CustomAsyncChildNodeSelector : IAsyncChildNodesSelector {
        private readonly IDataProvider dataProvider;

        public CustomAsyncChildNodeSelector(IDataProvider dataProvider) {
            this.dataProvider = dataProvider;
        }

        public Task<bool> HasChildNode(object item, CancellationToken token) {
            return System.Threading.Tasks.Task.Run(async () => {
                for(int i = 0; i < 10; i++) {
                    token.ThrowIfCancellationRequested();
                    await System.Threading.Tasks.Task.Delay(25);
                }
                return dataProvider.HasChildren(item);
            });
        }

        public IEnumerable SelectChildren(object item) {
            throw new NotImplementedException();
        }

        public Task<IEnumerable> SelectChildrenAsync(object item, CancellationToken token) {
            return System.Threading.Tasks.Task.Run(async () => {
                for(int i = 0; i < 20; i++) {
                    token.ThrowIfCancellationRequested();
                    await System.Threading.Tasks.Task.Delay(100);
                }
                return dataProvider.GetChildren(item);
            });
        }
    }

    public class FileSystemNodeImageSelector : TreeListNodeImageSelector {
        public override ImageSource Select(TreeListRowData rowData) {
            var fsItem = rowData.Row as FileSystemItem;
            if(fsItem == null) return null;

            if(fsItem.ItemType == "Drive")
                return FileSystemImages.DiskImage;
            else if(fsItem.ItemType == "Folder") {
                return rowData.IsExpanded ? FileSystemImages.OpenedFolderImage : FileSystemImages.ClosedFolderImage;
            }
            return FileSystemImages.FileImage;
        }
    }
}
